### PR TITLE
Make the documentation of completionItem larger

### PR DESCRIPTION
**Pull request by tamuratak**
This is a PR to make the documentation of completionItem larger, which solves #14165 along with commits by @sandy081 solving #67076.

Before:

After:

At various positions:

feb-25-2019 21-34-55

### CODE REVIEW

**Comment by tamuratak** [comment]
I found that this PR sovles #29126 too.

**Comment by Astrantia** [comment]
@sandy081 @joaomoreno Why not merge this for the next release?

File: src/vs/editor/contrib/suggest/suggestWidget.ts [hunk 1]
```diff
@@ -961,12 +959,95 @@ export class SuggestWidget implements IContentWidget, IListVirtualDelegate<Compl
...
+
+	private getDetailWidth(): number | null {
+		const editorWidth = this.getEditorWidth();
+		const leftDetailWidth = this.getLeftDetailWidth();
+		const rightDetailWidth = this.getRightDetailWidth();
+		if (leftDetailWidth === null || rightDetailWidth === null) {
+			return null;
+		}
+		const tmp = Math.max(leftDetailWidth, rightDetailWidth, this.listWidth);
+		return Math.min(tmp, editorWidth * 0.66);
```
Comment (joaomoreno) [inline src/vs/editor/contrib/suggest/suggestWidget.ts]: This should be rounded, since it will cause fractional numbers.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Looks good, but I would put this behind a setting and leave the default as is.

**Comment by tamuratak** [comment]
I would like to resolve the conflict if needed.

**Comment by joaomoreno** [comment]
Sure, go ahead :+1:

**Comment by joaomoreno** [comment]
Don't forget this:

> Looks good, **but I would put this behind a setting and leave the default as is.**

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Here are a few issues:

When there are no suggestions, this is what I see:

image

The code needs to react to the setting change. This is a bug:

1. Open suggest widget in one mode (small, for example)
2. Toggle the setting
3. Go back to the same editor, open suggest widget, it will still be in small mode

**Comment by jrieken** [comment]
Not a fan of the approach. This will only look good when there is loads of contents and this will look bad when there is little content. Can't we have something that adapts (within min/max bounds) to the size of the rendered contents?

**Comment by tamuratak** [comment]
Just like the solution in #69388, allow large details only when code blocks exist.

Aug-14-2019 08-52-18

**Review by jrieken — CHANGES_REQUESTED** [review]
Still not convinced of this solution, it should try to adopt to the real content size. The presence of a code block is some heuristic but I doubt it's good enough

**Comment by tamuratak** [comment]
With `vs/base/common/marked/marked`, the maximum length of lines in code blocks is used to set the width of details if they exist. If not, the width is set `500px` when this feature is enabled. When it is disabled, the width is set `330px`.

Enabled:





Disabled:



@joaomoreno, fixed reacting to the setting change. Fixed the size of `No suggestions`.

**Comment by MOZGIII** [comment]
This might not be perfect as it is, however it definitely solves a pain point if the screenshots represent the reality. What if this is merged in as is, and then polished to automatically adapt to the side alter? It looks like users can switch to this new mode only if they want to, so this should be good enough for now.

**Comment by awulkan** [comment]
This looks good enough to me for now. Right now the documentation window is barely readable in a lot of cases, and I'd hate to see a potential solution be stuck for years because it's not 100% perfect. It's better than what we currently have, and I believe it can be improved through incremental updates later on.

**Comment by wompeter** [comment]
These tiny popup windows have been a major usability issue for literally years, with people resorting to hacking the CSS and re-hacking it every update.  I can't believe the reluctance to deal with the problem when there are solutions like this waiting to go.

**Comment by jrieken** [comment]
I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

**Comment by tamuratak** [comment]
I heard it for the first time.

> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want.

Do you mean that you have started implementing a dynamic sizing suggestion widget? Can I try it on a certain branch?

> We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

**Comment by MOZGIII** [comment]
> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

Could you elaborate more on the desired approach? Or link us to a more in-depth discussion on the matter?

**Comment by tamuratak** [comment]
@jrieken  in the following gif, for example, are you saying that the width of details should shrink to the actual length of inside texts if they are short?

However, if the width of suggestions details changes too dynamically, it will annoy users.

Apr-15-2020 22-59-36

Moreover, when the width of suggestions changes dynamically, it is difficult to determine the correct position of it. In some cases, we have to change the position dynamically to prevent it going out of the window. And, it will annoy users.

**Comment by wompeter** [comment]
> I think we have been pretty clear that the approach of two static sizes (large and not so large) is not what we want. We are looking for dynamic sizing that tries to fit the contents of the suggestion details widget, actually we try to have the same for hovers as well.

Is anyone actually looking though?  Because this issue has been open in various forms since at least 2016.
https://github.com/microsoft/vscode/issues/14165

Debugging in vscode is so painful because of this.

**Pull request closed without merging**